sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java

```diff
@@ -448,4 +449,32 @@ private static Field convertField(TableFieldSchema bigQueryField) {
         bigQueryField.getDescription(),
         (Object) null /* Cast to avoid deprecated JsonNode constructor. */);
   }
+
+  private static Schema handleAvroLogicalTypes(TableFieldSchema bigQueryField, Type avroType) {
+    String bqType = bigQueryField.getType();
+    switch (bqType) {
+      case "NUMERIC":
+        // Default value based on
+        // https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#decimal_types
+        int precision = Optional.ofNullable(bigQueryField.getPrecision()).orElse(38L).intValue();
+        int scale = Optional.ofNullable(bigQueryField.getScale()).orElse(9L).intValue();
+        return LogicalTypes.decimal(precision, scale).addToSchema(Schema.create(Type.BYTES));
+      case "BIGNUMERIC":
+        // Default value based on
+        // https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#decimal_types
+        int precisionBigNumeric =
+            Optional.ofNullable(bigQueryField.getPrecision()).orElse(77L).intValue();
```

Comment:
The documentation you linked says `76.76` and the API docs say `1 <= precision - scale <= 38 and 0 <= scale <= 38` which means the max value for precision is `76`. I'm not sure what will work in practice it seems like `76` would potentially fail to read from BigQuery and `77` will fail to write to BigQuery. Can you provide any more info on why 77 was chosen? Should there be tests for this edge case?
https://cloud.google.com/java/docs/reference/google-cloud-bigquerystorage/2.8.2/com.google.cloud.bigquery.storage.v1.TableFieldSchema.Builder#com_google_cloud_bigquery_storage_v1_TableFieldSchema_Builder_getPrecision__

Reply:
I ran a beam job and tried reading from an actual BigQuery table with BIGNUMERIC field. The [writer schema](https://github.com/apache/beam/blob/483a0c95734c528aa45419596e9f27e9e650c5d7/sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroSource.java#L754), set in the extracted Avro file, has precision value set as 77 and the scale is set to 38, for fields that have no explicit precision/scale specified.  

I also tried inserting the max supported value that has a precision of 77 and it works as long as its <= 578960446186580977117854925043439539266.34992332820282019728792003956564819967 (since 77th digit is partial it supports this value and not 999999999999999999999999999999999999999.99999999999999999999999999999999999999 instead)

Reply:
The tests you describe don't seem to be in the code and this didn't answer my question but I guess the answer is in the initial question. A value of `77` will at least pull the failure into something the user can fix with additional constraints within their pipeline.

Reply:
sorry, i was not referring to these unit tests. I tested with a standalone beam job, in direct runner mode, against an actual BigQuery table and verified the default precision/scale value set in the Avro writer schema (that's why chose to set default as 77 instead of 76). Also, tried inserting and reading the max documented value for BigNumeric field and it works for numbers with 77 precision. I could add a BQ integration test to test this edge case, if that helps.